File: hello-php/src/Server/Protocol.php
<?php

declare(strict_types=1);

namespace AudioStream\Server;

enum CommandType: string
{
    case STREAM = 'STREAM';
    case DATA = 'DATA';
    case QUERY = 'QUERY';
}

enum StreamCommand: string
{
    case CREATE = 'CREATE';
    case COMPLETE = 'COMPLETE';
    case CLOSE = 'CLOSE';
}

enum DataCommand: string
{
    case READ = 'READ';
}

enum QueryCommand: string
{
    case GET_STATUS = 'GET_STATUS';
    case LIST_STREAMS = 'LIST_STREAMS';
}

final class CommandInfo
{
    public function __construct(
        public readonly CommandType $type,
        public readonly string $command,
    ) {}
}

final class Protocol
{
    /** @var array<string, CommandInfo> */
    private static array $map = [];

    public static function lookup(string $command): ?CommandInfo
    {
        if (empty(self::$map)) {
            foreach (StreamCommand::cases() as $c) {
                self::$map[$c->value] = new CommandInfo(CommandType::STREAM, $c->value);
            }
            foreach (DataCommand::cases() as $c) {
                self::$map[$c->value] = new CommandInfo(CommandType::DATA, $c->value);
            }
            foreach (QueryCommand::cases() as $c) {
                self::$map[$c->value] = new CommandInfo(CommandType::QUERY, $c->value);
            }
        }
        return self::$map[$command] ?? null;
    }
}

File: hello-dart/lib/server/protocol.dart
enum CommandType { stream, data, query }

enum StreamCommand {
  create('CREATE'),
  complete('COMPLETE'),
  close('CLOSE');

  final String value;
  const StreamCommand(this.value);
}

enum DataCommand {
  read('READ');

  final String value;
  const DataCommand(this.value);
}

enum QueryCommand {
  getStatus('GET_STATUS'),
  listStreams('LIST_STREAMS');

  final String value;
  const QueryCommand(this.value);
}

class CommandInfo {
  final CommandType type;
  final String command;
  const CommandInfo(this.type, this.command);
}

class Protocol {
  static final Map<String, CommandInfo> _map = _buildMap();

  static Map<String, CommandInfo> _buildMap() {
    final m = <String, CommandInfo>{};
    for (final c in StreamCommand.values) {
      m[c.value] = CommandInfo(CommandType.stream, c.value);
    }
    for (final c in DataCommand.values) {
      m[c.value] = CommandInfo(CommandType.data, c.value);
    }
    for (final c in QueryCommand.values) {
      m[c.value] = CommandInfo(CommandType.query, c.value);
    }
    return m;
  }

  static CommandInfo? lookup(String command) => _map[command];
}
